develop/reference_processing.rst

```diff
@@ -34,31 +34,18 @@ Distributed Query Executor
 
 Citus’s distributed executors run distributed query plans and handle failures that occur during query execution. The executors connect to the workers, send the assigned tasks to them and oversee their execution. If the executor cannot assign a task to the designated worker or if a task execution fails, then the executor dynamically re-assigns the task to replicas on other workers. The executor processes only the failed query sub-tree, and not the entire query while handling failures.
 
-Citus has three basic executor types: real time, router, and task tracker. It chooses which to use dynamically, depending on the structure of each query, and can use more than one at once for a single query, assigning different executors to different subqueries/CTEs as needed to support the SQL functionality. This process is recursive: if Citus cannot determine how to run a subquery then it examines sub-subqueries.
+Citus has three executor types: adaptive, push-pull, and task tracker. It chooses which to use dynamically, depending on the structure of each query, and can use more than one at once for a single query, assigning different executors to different subqueries/CTEs as needed to support the SQL functionality. This process is recursive: if Citus cannot determine how to run a subquery then it examines sub-subqueries.
```

Comment:
I'm not sure whether we should count `push-pull` as a separate executor.

I don't know how to express this well, but I can explain to you and you can probably come-up with a better explanation of it.

`pull-push` executor is actually not a different executor. It is calling the current executor to execute `some parts` of the query like ctes or some subqueries. 

For example,

```SQL
WITH distinct_users AS (SELECT DISTINCT user_id FROM users_table)
SELECT count(*) FROM events_table WHERE user_id IN (SELECT user_id FROM distinct_users);
```

For example, for the above query, the execution happens in the following order:

1)  The query in `distinct_users` CTE is executed via adaptive executor as if it is a completely separate query. 
2) The results of the  `distinct_users` CTE is first pulled to the coordinator as usual query execution, and copied (`pushed`) to the workers as intermediate result. 
3) The final query `SELECT count(*) FROM events_table WHERE user_id IN (SELECT user_id FROM read_intermediate_result('9_1'::text, 'binary'::citus_copy_format) intermediate_result(user_id integer)) distinct_users)` is executed by adaptive executor. Basically, treating the intermediate result as a reference table, knowing that it exits on all worker nodes.

So, in that sense, I'd probably frame `pull-push` execution as a `tool` that adaptive executor (not task-tracker!) can use to execute more complex SQL that otherwise cannot be executed by Citus.

But, if framing it as a separate executor would make it easier to explain to the users, that's also not too bad.